compiler/block.ts

```diff
@@ -0,0 +1,181 @@
+import * as assert from 'node:assert';
+
+import { RuntimeError } from '../exceptions';
+import { RuntimeFault } from '../faults';
+import {
+  Instr,
+  InstrVisitor,
+  Print,
+  Expression,
+  Rem,
+  New,
+  Load,
+  List,
+  Renum,
+  Save,
+  Run,
+  Exit,
+  End,
+  Let,
+  Assign,
+  ShortIf,
+  If,
+  Else,
+  ElseIf,
+  EndIf,
+} from '../ast/instr';
+
+import { LineCompiler } from './base';
+
+//
+// A class to manage blocks in the LineCompiler. Subclassed for particular
+// kinds of blocks in the base, next to the LineCompiler.
+//
+
+// NOTE: Conceptually, BlockKind is enumerable. However, as the kinds are
+// for the most part defined in the base, setting this to a string allows us
+// to avoid a generic type. In practice, block kinds are only used for error
+// reporting.
+export type BlockKind = string;
+
+export abstract class Block implements InstrVisitor<void> {
+  public kind: BlockKind = '<unknown>';
+  private _compiler: LineCompiler | null = null;
+  public instr: Instr | null = null;
+  public previous: Block | null = null;
+  public parent: Block | null = null;
+
+  public init(
+    compiler: LineCompiler,
+    instr: Instr | null,
+    previous: Block | null,
+    parent: Block | null,
+  ) {
+    this._compiler = compiler;
+    this.instr = instr;
+    this.previous = previous;
+    this.parent = parent;
+  }
+
+  public get compiler(): LineCompiler {
+    if (!this._compiler) {
+      throw RuntimeFault.fromError(
+        new RuntimeError(`Block ${this.kind} not initialized`),
+      );
+    }
+    return this._compiler;
+  }
+
+  // Open a new block, maintaining a reference to the parent block.
+  public begin(instr: Instr, block: Block): void {
+    block.init(this.compiler, instr, null, this.compiler.block);
+    this.compiler.block = block;
+  }
+
+  // Go to the next block in a chain of blocks
+  public next(instr: Instr, block: Block): void {
+    block.init(
+      this.compiler,
+      instr,
+      this.compiler.block,
+      this.compiler.block.parent,
+    );
+    this.compiler.block = block;
+  }
+
+  // End the current block.
+  public end(): void {
+    assert.ok(this.parent);
+    this.compiler.block = this.parent;
+  }
+
+  public handle(instr: Instr): void {
+    instr.accept(this);
+  }
+
+  // Most instructions do not end any kind of block. This code should never
+  // be called by the compiler.
+  // TODO: Shouldn't "never" work for the return type, since syntaxFault has
+  // a return type of never?
+  private invalid(instr: Instr, name: string): void {
+    this.compiler.syntaxFault(instr, `${name} can not end blocks`);
+  }
+
+  // Some instructions end specific kinds of blocks. By default, such an
+  // instruction does not close *this* kind of block.
+  public mismatched(instr: Instr, kind: string): void {
+    this.compiler.syntaxFault(instr, `${kind} can not end ${this.kind}`);
+  }
+
+  visitPrintInstr(print: Print): void {
```

Comment:
It's going to be a little tedious, but let's write the tests for invalid and mismatched. It will be worth the coverage.